source/distributed/tria.cc

```diff
@@ -1423,27 +1626,126 @@ namespace parallel
+                               offset_variable_data_sizes +
+                               callback_index * sizeof(unsigned int);
+
+          // Let the data iterator point to the correct buffer.
+          dest_data_it = dest_data_variable.cbegin();
+        }
+      else
+        {
+          // For the fixed size data, we can get the information about
+          // the buffer location on each cell directly from the
+          // sizes_fixed_cumulative vector.
+          offset         = sizes_fixed_cumulative[callback_index];
+          size           = sizes_fixed_cumulative[callback_index + 1] - offset;
+          data_increment = sizes_fixed_cumulative.back();
+
+          // Let the data iterator point to the correct buffer.
+          // Adjust buffer iterator to the offset of the callback
+          // function so that we only have to advance its position
+          // to the next cell after each iteration.
+          dest_data_it = dest_data_fixed.cbegin() + offset;
+        }
 
       // Iterate over all cells and unpack the transferred data.
-      // Adjust buffer iterator to the offset of the callback
-      // function and advance its position to the next cell
-      // after each iteration.
       auto quad_cell_rel_it = quad_cell_relations.begin();
-      auto dest_fixed_it    = dest_data_fixed.cbegin() + offset;
+      auto dest_sizes_it    = dest_sizes_variable.cbegin();
       for (; quad_cell_rel_it != quad_cell_relations.end();
-           ++quad_cell_rel_it, dest_fixed_it += cumulative_sizes_fixed.back())
+           ++quad_cell_rel_it, dest_data_it += data_increment)
         {
           const auto &cell_status = std::get<1>(*quad_cell_rel_it);
           const auto &dealii_cell = std::get<2>(*quad_cell_rel_it);
 
+          if (callback_variable_transfer)
+            {
+              // Update the increment according to the whole data size
+              // of the current cell.
+              data_increment = *dest_sizes_it;
+
+              if (cell_status !=
+                  parallel::distributed::Triangulation<dim,
+                                                       spacedim>::CELL_INVALID)
+                {
+                  // Extract the corresponding values for offset and size from
+                  // the cumulative sizes array stored in the fixed size buffer.
+                  offset = 0;
+                  if (callback_index == 0)
+                    {
+                      size = Utilities::unpack<unsigned int>(
+                        dest_sizes_cell_it,
+                        dest_sizes_cell_it + sizeof(unsigned int),
+                        /*allow_compression=*/false);
+                    }
+                  else
+                    {
+                      offset = Utilities::unpack<unsigned int>(
+                        dest_sizes_cell_it,
+                        dest_sizes_cell_it + sizeof(unsigned int),
+                        /*allow_compression=*/false);
```

Comment:
this here (and the next statement) would be the place where you'd use `memcpy` to get the data back